x/poolmanager/routes.go

```diff
@@ -0,0 +1,426 @@
+	"github.com/osmosis-labs/osmosis/v21/x/poolmanager/types"
+)
+
+var (
+	// We cache direct routes and spot prices to avoid recalculating them.
+	// It is important to note, these cache values are only used within the same query.
+	// If a new query is made, the cache will be reset.
+	directRouteCache map[string]uint64
+	spotPriceCache   map[string]osmomath.BigDec
+	shouldCache      = false
+)
+
+func init() {
+	directRouteCache = make(map[string]uint64)
+	spotPriceCache = make(map[string]osmomath.BigDec)
+}
+
+// findRoutes finds all routes between two tokens that match the given hop count
+func findRoutes(g types.RoutingGraphMap, start, end string, hops int) [][]*types.Route {
+	if hops < 1 {
+		return nil
+	}
+
+	var routeRoutes [][]*types.Route
+
+	startRoutes, startExists := g.Graph[start]
+	if !startExists {
+		return routeRoutes
+	}
+
+	for token, routes := range startRoutes.InnerMap {
+		if hops == 1 {
+			if token == end {
+				for _, route := range routes.Routes {
+					route.Token = end
+					routeRoutes = append(routeRoutes, []*types.Route{route})
+				}
+			}
+		} else {
+			subRoutes := findRoutes(g, token, end, hops-1)
+			for _, subRoute := range subRoutes {
+				for _, route := range routes.Routes {
+					route.Token = token
+					fullRoute := append([]*types.Route{route}, subRoute...)
+					routeRoutes = append(routeRoutes, fullRoute)
+				}
+			}
+		}
+	}
+
+	return routeRoutes
+}
+
+// SetDenomPairRoutes sets the route map to be used for route calculations.
+func (k Keeper) SetDenomPairRoutes(ctx sdk.Context) (types.RoutingGraph, error) {
+	// Reset cache at the end of this function
+	// We only ever want to use cache in the helper functions if it is being called via this function,
+	// since this function is called at upgrade and epoch at determinstic times. If we used cache outside of this function,
+	// it's possible the cache will mess with the determinism of the route map.
+	shouldCache = true
+	defer func() {
+		directRouteCache = make(map[string]uint64)
+		spotPriceCache = make(map[string]osmomath.BigDec)
+		shouldCache = false
+	}()
+
+	// We generate all denom pair routes here, and use them to filter out pools that do not meet the minimum liquidity threshold.
+	// While generating routes twice is not ideal, we are optimizing to reduce writes to state.
+	pools, routeMap, err := k.generateAllDenomPairRoutes(ctx)
+	if err != nil {
+		return types.RoutingGraph{}, err
+	}
+
+	// Retrieve minimum liquidity threshold from params
+	minValueForRoute := k.GetParams(ctx).MinValueForRoute
+
+	// Create a routingGraph to represent possible routes between tokens
+	var routingGraph types.RoutingGraph
+
+PoolLoop:
+	// Iterate through the pools
+	for _, pool := range pools {
+		// Some of the first cw pools created have a malformed response and are no longer in use. Remove these pools to prevent issues.
+		tokens := pool.GetPoolDenoms(ctx)
+		for _, token := range tokens {
+			if strings.Contains(token, "pool_asset_denoms") {
```

Comment:
The first 3 CW pools that were created have a malformed response when querying the denoms it consists of (it's response is a full json, and it has zero tokens since it is broken). If a denom comes around that contains the exact words "pool_asset_denoms"...then they are going to just have to take the L on not being included in the on chain router for having a dumb name lol